runelite-client/src/main/java/net/runelite/client/game/chatbox/ChatboxTextInput.java

```diff
@@ -0,0 +1,663 @@
+		promptWidget.setOriginalY(8);
+		promptWidget.setOriginalHeight(24);
+		promptWidget.setXTextAlignment(WidgetTextAlignment.CENTER);
+		promptWidget.setYTextAlignment(WidgetTextAlignment.CENTER);
+		promptWidget.setWidthMode(WidgetSizeMode.MINUS);
+		promptWidget.revalidate();
+
+		buildEdit(0, 50, container.getWidth(), 0);
+	}
+
+	protected void buildEdit(int x, int y, int w, int h)
+	{
+		Widget container = chatboxPanelManager.getContainerWidget();
+
+		String lt = value.substring(0, this.cursor);
+		String mt = value.substring(this.cursor, this.cursorEnd);
+		String rt = value.substring(this.cursorEnd);
+
+		Widget leftText = container.createChild(-1, WidgetType.TEXT);
+		Widget cursor = container.createChild(-1, WidgetType.RECTANGLE);
+		Widget middleText = container.createChild(-1, WidgetType.TEXT);
+		Widget rightText = container.createChild(-1, WidgetType.TEXT);
+
+		leftText.setFontId(fontID);
+		FontTypeFace font = leftText.getFont();
+
+		if (h <= 0)
+		{
+			h = font.getBaseline();
+		}
+
+		int ltw = font.getTextWidth(lt);
+		int mtw = font.getTextWidth(mt);
+		int rtw = font.getTextWidth(rt);
+
+		int fullWidth = ltw + mtw + rtw;
+
+		int ox = x;
+		if (w > 0)
+		{
+			x += (w - fullWidth) / 2;
+		}
+
+		int ltx = x;
+		int mtx = ltx + ltw;
+		int rtx = mtx + mtw;
+
+		leftText.setText(lt);
+		leftText.setOriginalX(ltx);
+		leftText.setOriginalY(y);
+		leftText.setOriginalWidth(ltw);
+		leftText.setOriginalHeight(h);
+		leftText.revalidate();
+
+		if (mt.length() > 0)
+		{
+			cursor.setTextColor(0x113399);
+		}
+		else
+		{
+			cursor.setTextColor(0xFFFFFF);
+			cursor.setOnTimerListener((JavaScriptCallback) ev ->
+			{
+				boolean on = System.currentTimeMillis() % CURSOR_FLASH_RATE_MILLIS > (CURSOR_FLASH_RATE_MILLIS / 2);
+				cursor.setOpacity(on ? 255 : 0);
+			});
+		}
+		cursor.setFilled(true);
+		cursor.setOriginalX(mtx - 1);
+		cursor.setOriginalY(y);
+		cursor.setOriginalWidth(2 + mtw);
+		cursor.setOriginalHeight(h);
+		cursor.revalidate();
+
+		middleText.setText(mt);
+		middleText.setFontId(fontID);
+		middleText.setOriginalX(mtx);
+		middleText.setOriginalY(y);
+		middleText.setOriginalWidth(mtw);
+		middleText.setOriginalHeight(h);
+		middleText.setTextColor(0xFFFFFF);
+		middleText.revalidate();
+
+		rightText.setText(rt);
+		rightText.setFontId(fontID);
+		rightText.setOriginalX(rtx);
+		rightText.setOriginalY(y);
+		rightText.setOriginalWidth(rtw);
+		rightText.setOriginalHeight(h);
+		rightText.revalidate();
+
+		net.runelite.api.Point ccl = container.getCanvasLocation();
+		int canvasX = ltx + ccl.getX();
+		Rectangle bounds = new Rectangle(ccl.getX() + ox, ccl.getY() + y, w > 0 ? w : fullWidth, h);
+
+		String tsValue = value.toString();
+		isInBounds = ev -> bounds.contains(ev.getPoint());
+		getCharOffset = ev ->
+		{
+			int cx = ev.getX() - canvasX;
+
+			int c = (tsValue.length() * cx) / fullWidth;
+
+			for (; c >= 0 && c <= tsValue.length(); )
```

Comment:
Is there a rationale for this over a while loop? Besides making me think I'm going insane whilst reviewing it

Reply:
I just always use `for` loops. Its a Go thing.

Reply:
It feels dirty, whilst I recognised the construct,  you're bastardising a for `loop` into a `while` even though the language has a `while` implementation.

Functionality wise, they're identical, but for the sake of the minds of sweet innocent java developers(and future maintainability) wouldn't it be better to use a while.

Reply:
I agree with Nathen, this is not Go and you should use actual Java language features even though they are missing in `Go`

Reply:
Well you are in luck, I realized there is an edge case so its a real for loop now 🤷‍♂️